### PR TITLE
[WIP] socat on a node instead nsenter in a case of rkt-kvm driver

### PR DESCRIPTION
**Pull request by borqosky**

### CODE REVIEW

**Comment by woodbor** [comment]
cc @yifan-gu

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -1887,15 +1886,27 @@ func (r *Runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
...
 
-	command := exec.Command(nsenterPath, args...)
+	var args []string
+	if r.config.Stage1Image == "stage1-kvm.aci" {
+		args = append(args, "enter", fmt.Sprintf("--app=%s", listResp.Pods[0].Apps[0].Name), listResp.Pods[0].Id)
+	} else {
+		args = append(args, "-t", fmt.Sprintf("%d", listResp.Pods[0].Pid), "-n")
+	}
+
+	args = append(args, socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port))
```
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: What if the socat does not exist inside the kvm?

**Comment by yifan-gu** [comment]
@woodbor Sorry for the delay, was still in china till sunday.

Have you tested this with the e2e tests? See https://github.com/kubernetes/kubernetes/issues/25117#issuecomment-221443207

**Comment by pskrzyns** [comment]
@yifan-gu if i understand correctly, port forwarding is used only for debugging and it should works only on node where pod is running, so why not leave main logic implementation on the rkt side and use kubelet just for grpc call to rkt api-service ?

**Comment by yifan-gu** [comment]
@pskrzyns Even if we put portforward logic on the rkt side, we will need to implement for all flavors either.
Though I am not arguing kubelet is the best place to implement portforward. cc @yujuhong 

But how do we solve the issue that the container might not have `socat` ? Copy the binary into the stage1 or making `rkt enter` able to copy and run the binary into the container?

File: pkg/kubelet/rkt/rkt.go [hunk 2]
```diff
@@ -1887,15 +1886,29 @@ func (r *Runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
...
+		args = append(args, "-t", fmt.Sprintf("%d", listResp.Pods[0].Pid), "-n")
+	}
+
+	if r.config.Stage1Image == "stage1-kvm.aci" {
+		args = append(args, "-", fmt.Sprintf("TCP4:%s:%d", listResp.Pods[0].GetNetworks()[1].Ipv4, port))
+	} else {
+		args = append(args, socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port))
+	}
+
+	command := new(exec.Cmd)
```
Comment (jellonek) [inline pkg/kubelet/rkt/rkt.go]: move this line just after `var args` and then combine all 3 ifs into single one.

**Comment by pskrzyns** [comment]
@yifan-gu can you close this pull request ? It's not valid anymore

**Comment by yifan-gu** [comment]
From OOB discussion with @pskrzyns , we will try to have rkt api service implement the portforward using iptables.

**Pull request closed without merging**